src/ota_proxy/utils.py

```diff
@@ -1,22 +1,70 @@
 from __future__ import annotations
 
+import os
 from hashlib import sha256
 from os import PathLike
-from typing import AsyncIterator
+from typing import AsyncGenerator
+from urllib.parse import SplitResult, quote, urlsplit
 
+import anyio
 from anyio import open_file
 
+from otaclient_common._typing import StrOrPath
+
 from .config import config as cfg
 
 
-async def read_file(fpath: PathLike) -> AsyncIterator[bytes]:
+async def read_file(
+    fpath: PathLike, chunk_size: int = cfg.LOCAL_READ_SIZE
+) -> AsyncGenerator[bytes]:
     """Open and read a file asynchronously."""
     async with await open_file(fpath, "rb") as f:
-        while data := await f.read(cfg.CHUNK_SIZE):
+        fd = f.wrapped.fileno()
+        os.posix_fadvise(fd, 0, 0, os.POSIX_FADV_SEQUENTIAL)
+        while data := await f.read(chunk_size):
             yield data
+        os.posix_fadvise(fd, 0, 0, os.POSIX_FADV_DONTNEED)
+
+
+def read_file_once(fpath: StrOrPath | anyio.Path) -> bytes:
+    with open(fpath, "rb") as f:
+        fd = f.fileno()
+        os.posix_fadvise(fd, 0, 0, os.POSIX_FADV_SEQUENTIAL)
+        data = f.read()
+        os.posix_fadvise(fd, 0, 0, os.POSIX_FADV_DONTNEED)
+    return data
```

Comment:
Use `os.posix_fadvice` to prevent kernel holds the open/written files page caches after files are closed.